dd-sdk-android/src/main/kotlin/com/datadog/android/core/internal/data/batching/DefaultConsentAwareDataHandler.kt

```diff
@@ -0,0 +1,147 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2016-Present Datadog, Inc.
+ */
+
+package com.datadog.android.core.internal.data.batching
+
+import com.datadog.android.core.internal.data.batching.migrators.BatchedDataMigrator
+import com.datadog.android.core.internal.data.batching.processors.DataProcessor
+import com.datadog.android.core.internal.data.privacy.Consent
+import com.datadog.android.core.internal.data.privacy.ConsentProvider
+import com.datadog.android.core.internal.data.privacy.ConsentProviderCallback
+
+internal class DefaultConsentAwareDataHandler<T : Any>(
+
+    private val consentProvider: ConsentProvider,
+    private val permissionGrantedProcessorFactory:
+        (BatchedDataMigrator, ConsentProvider) -> DataProcessor<T>,
+    private val permissionPendingProcessorFactory:
+        (BatchedDataMigrator, ConsentProvider) -> DataProcessor<T>,
+    private val noOpProcessorFactory: () -> DataProcessor<T>,
+    private val pendingToGrantedMigratorFactory: () -> BatchedDataMigrator,
+    private val pendingToNotGrantedMigratorFactory: () -> BatchedDataMigrator,
+    private val pendingToPendingMigratorFactory: () -> BatchedDataMigrator,
+    private val grantedToGrantedMigratorFactory: () -> BatchedDataMigrator,
+    private val grantedToNotGrantedMigratorFactory: () -> BatchedDataMigrator,
+    private val grantedToPendingMigratorFactory: () -> BatchedDataMigrator,
+    private val notGrantedToGrantedMigratorFactory: () -> BatchedDataMigrator,
+    private val notGrantedToNotGrantedMigratorFactory: () -> BatchedDataMigrator,
+    private val notGrantedToPendingMigratorFactory: () -> BatchedDataMigrator
```

Comment:
We don't need to consider those transitions:
```kotlin
private val pendingToPendingMigratorFactory: () -> BatchedDataMigrator,
private val grantedToGrantedMigratorFactory: () -> BatchedDataMigrator,
private val notGrantedToNotGrantedMigratorFactory: () -> BatchedDataMigrator,
```
if we ensure that `ConsentProvider` does not notify on no-state change. This is much easier to test in `provider` than `handler`. WDYT?

Reply:
Yeah, that's a good point actually, will add the changes ;)

Reply:
As a matter a fact we need those, I actually thought of this when I added the logic and the reason why I added them is that first time you will initialise the DataHandler you will need also a processor/migrator as you need to do something with the data that was from last usage. In this case it is something like pending->pending, granted->granted or notgranted->notgranted following our logic. Again these cases are needed for the moment when you initialise the Handler. 
Regarding ConentProvider I could add this optimisation but I am thinking if this is the right thing to do, basically swallowing some updates coming from the client just because the value was the same. Maybe later we need to react on any update no matter the previous values and this will create issues.
@xgouchet @buranmert What do you think ?